include/execution.hpp

```diff
@@ -2171,7 +2172,257 @@ namespace std::execution {
+          }
+        }
+
+        template <class... _As>
+          requires (!tag_invocable<set_value_t, _Receiver, decay_t<_As>&...>)
+        void __apply_values(tuple<_As...>&) noexcept {
+          assert(!"_Should never get here");
+        }
+
+        void __propagate_signal() noexcept {
+          auto &__data = __shared_state_->__data_;
+
+          if (__data.index() == 0) {
+            visit([&](auto& __tupl) -> void {
+              __apply_values(__tupl);
+            }, std::get<0>(__data));
+          }
+          if (__data.index() == 1) {
+            visit([&](auto& __err) -> void {
+              execution::set_error((_Receiver&&) __recvr_, __err);
+            }, std::get<1>(__data));
+          }
+          else if (__data.index() == 2) {
+            execution::set_stopped((_Receiver&&) __recvr_);
+          }
+        }
+
+      public:
+        __operation(_Receiver&& __rcvr,
+                    shared_ptr<__sh_state<_Sender>> __shared_state)
+          : __recvr_((_Receiver&&)__rcvr)
+          , __shared_state_(move(__shared_state)) {
+        }
+
+        void __notify() noexcept override {
+          __on_stop_.reset();
+          __propagate_signal();
+        }
+
+        friend void tag_invoke(start_t, __operation& __self) noexcept try {
+          __sh_state<_Sender> *__shared_state = __self.__shared_state_.get();
+          unique_lock __lock{__shared_state->__mutex_};
+          __state_t __state = __shared_state->__state_;
+
+          if (__state == __state_t::__completed) {
+            __lock.unlock();
+            __self.__propagate_signal();
+          }
+          else {
+            __self.__on_stop_.emplace(
+                get_stop_token(get_env(__self.__recvr_)),
+                __on_stop_requested{__shared_state->__stop_source_});
+
+            if (__shared_state->__stop_source_.stop_requested()) {
+              __lock.unlock();
+              execution::set_stopped((_Receiver&&) __self.__recvr_);
+            }
+            else {
+              __shared_state->__operation_states_.push_back(&__self);
+
+              if (__state != __state_t::__started) {
+                __shared_state->__state_ = __state_t::__started;
+                __lock.unlock();
+
+                execution::start(__shared_state->__op_state2_);
+              }
+            }
+          }
+        } catch (...) {
+          execution::set_error((_Receiver&&) __self.__recvr_, current_exception());
+        }
+      };
+
+    template <class _SenderId>
+      class __sender {
+        using _Sender = __t<_SenderId>;
+        using __sh_state = __sh_state<_Sender>;
+        template <class _Receiver>
+          using __operation = __operation<_SenderId, __x<remove_cvref_t<_Receiver>>>;
+
+        _Sender __sndr_;
+        shared_ptr<__sh_state> __shared_state_;
+
+      public:
+        template <__decays_to<__sender> _Self, receiver _Receiver>
```

Comment:
Thanks for spotting this! I've forgotten to constrain the connect method. Although it's possible to use the constrain you've mentioned, it won't reduce the overload set (since `_receiver` doesn't depend on `_Receiver` in any way). Also, the completion signatures of `__member_t<_Self, _Sender>` and `_Self` are quite different. If you don't mind, I'll use a bit different constrain here:
```
receiver_of<_Receiver, completion_signatures_of_t<_Self>>
```

Reply:
I think this needs to be at least:

```c++
receiver_of<
  _Receiver,
  completion_signatures_of_t<_Self, __empty_env>>
```

... along with the change below to the customization of `get_completion_signatures`.